# 🍀 로또 게임

## ✅ 기능 구현 목록

- [x] `구입 금액을 입력 받는다.`

  - [x] (예외처리) 1,000원 단위로 나누어 떨어져야한다.
  - [x] (예외처리) 입력은 숫자로만 이루어져야 한다.

- [x] `당첨 번호를 입력 받는다.`

  - [x] (예외처리) 입력은 숫자로만 이루어져야 한다.
  - [x] (예외처리) 1부터 45까지의 범위여야한다.
  - [x] (예외처리) 6개의 숫자에 중복이 없어야한다.

- [x] `보너스 번호를 입력 받는다.`

  - [x] (예외처리) 입력은 숫자로만 이루어져야 한다.
  - [x] (예외처리) 1부터 45까지의 범위여야한다.
  - [x] (예외처리) 당첨 번호와 중복되는 숫자가 아니어야한다.

- [x] `구입 금액 입력에 따라 몇 개의 로또를 샀는지 판단한다.`

- [x] `1부터 45까지의 서로 다른 임의의 수 6개를 구입한 로또의 수만큼 생성한다.`

  - [x] 랜덤한 6개의 로또 번호 생성 기능
  - [x] (예외처리) 로또 숫자의 갯수가 6개여야 한다.
  - [x] (예외처리) 6개의 숫자에 중복이 없어야한다.
  - [x] 생성된 로또 번호를 오름차순으로 정렬한다.

- [x] `각 로또의 임의로 생성된 수와 당첨 번호를 비교하여 결과를 반환한다.`

  - [x] 생성된 로또 번호를 각각 뽑아낸다.
  - [x] 뽑아낸 로또 번호를 입력된 당첨 번호와 비교한다.
  - [x] 뽑아낸 로또 번호에 입력된 보너스 번호가 포함되어 있는지 확인한다.
  - [x] 번호 비교를 통해 로또 등수를 알아낸다.
  - [x] 등수를 종합한다.
  - [x] 당첨 금액을 계산하여 반환한다.
  - [x] 수익률을 계산하고 소수점 둘째 자리에서 반올림한 값을 반환한다.

- [x] `반환된 결과에 따라 당첨 내역을 출력한다.`

- [x] `수익률을 출력한다.`

- [x] `예외 상황 시 [ERROR]로 시작하는 에러 문구를 출력한다.`

### 🛠 테스트

- [x] `입력에 대한 유효성 테스트 (InputValidationTest)`

  - [x] 구입 금액이 로또 금액으로 나누어 떨어지는지에 대한 메서드 테스트 (Validation.isDivided)
  - [x] 구입 금액이 모두 숫자인지 검사하는 메서드 테스트 (Validation.isOnlyNumber)
  - [x] 입력된 당첨 번호 각 숫자들의 범위가 1이상 45이하인지 검사하는 메서드 테스트 (Validation.isNumberRange)
  - [x] 입력된 보너스 번호의 숫자가 당첨 번호 숫자와 중복되는지 검사하는 메서드 테스트 (Validation.isUniqueBonusNumber)

- [x] `ApplicationTest`

- [x] `LottoTest`

  - [x] 로또 번호 입력이 공백이면 예외가 발생하는지 테스트
  - [x] 로또 번호 입력에 문자가 있으면 예외가 발생하는지 테스트
  - [x] 로또 번호 입력이 지정된 범위를 벗어나면 예외가 발생하는지 테스트

## 🚀 기능 요구 사항

로또 게임 기능을 구현해야 한다. 로또 게임은 아래와 같은 규칙으로 진행된다.

```
- 로또 번호의 숫자 범위는 1~45까지이다.
- 1개의 로또를 발행할 때 중복되지 않는 6개의 숫자를 뽑는다.
- 당첨 번호 추첨 시 중복되지 않는 숫자 6개와 보너스 번호 1개를 뽑는다.
- 당첨은 1등부터 5등까지 있다. 당첨 기준과 금액은 아래와 같다.
    - 1등: 6개 번호 일치 / 2,000,000,000원
    - 2등: 5개 번호 + 보너스 번호 일치 / 30,000,000원
    - 3등: 5개 번호 일치 / 1,500,000원
    - 4등: 4개 번호 일치 / 50,000원
    - 5등: 3개 번호 일치 / 5,000원
```

- 로또 구입 금액을 입력하면 구입 금액에 해당하는 만큼 로또를 발행해야 한다.
- 로또 1장의 가격은 1,000원이다.
- 당첨 번호와 보너스 번호를 입력받는다.
- 사용자가 구매한 로또 번호와 당첨 번호를 비교하여 당첨 내역 및 수익률을 출력하고 로또 게임을 종료한다.
- 사용자가 잘못된 값을 입력할 경우 `throw`문을 사용해 예외를 발생시키고, "[ERROR]"로 시작하는 에러 메시지를 출력 후 종료한다.
